src/imputation/MoR.py

```diff
@@ -112,8 +125,148 @@ def carry_forwards(df, backdata, impute_vars):
         df.loc[match_cond, f"{var}_imputed"] = df.loc[match_cond, f"{var}_prev"]
     df.loc[match_cond, "imp_marker"] = "CF"
 
+    df.loc[match_cond] = create_imp_class_col(
+        df, "200_prev", "201_prev", use_cellno=False
+    )
+
     # Drop merge related columns
-    to_drop = [column for column in df.columns if column.endswith("_prev")]
+    to_drop = [
+        column
+        for column in df.columns
+        if (column.endswith("_prev"))
+        & (re.search("(.*)_prev|.*", column).group(1) not in impute_vars)
+    ]
     to_drop += ["_merge"]
     df = df.drop(to_drop, axis=1)
     return df
+
+
+def calculate_growth_rates(current_df, prev_df, target_vars):
+    """Calculate the growth rates between previous and current data.
+
+    Args:
+        current_df (pd.DataFrame): pre-processed current data.
+        prev_df (pd.DataFrame): pre-processed backdata.
+        target_vars ([string]): target vars to impute.
+    """
+    # Only calculate links for long form responders
+    current_df = current_df.copy().loc[current_df["formtype"] == "0001", :]
+    # prev_df = prev_df.copy().loc[prev_df["formtype"] == "0001", :]
+
+    gr_df = pd.merge(
+        current_df,
+        prev_df,
+        on=["reference", "imp_class"],
+        how="inner",
+        suffixes=("", "_prev"),
+    )
+    target_vars_prev = [f"{var}_prev" for var in target_vars]
+    gr_df = (
+        gr_df[["reference", "imp_class"] + target_vars + target_vars_prev]
+        .groupby(["reference", "imp_class"])
+        .sum()
+    ).reset_index()
```

Comment:
again, I think you might need drop=True in here

Reply:
I did remove this but realised it broke the code as the index here is "imp_class" which we need back in as a column.

Reply:
I didn't realise it worked like that. Cool.